### ISSUE TITLE
Support for samurai (alternate ninja implementation)

### ISSUE THREAD

**Issue by michaelforney** [mesonbuild/meson#3405]
I'm the author of [samurai](https://github.com/michaelforney/samurai), an alternative implementation of ninja, and I'd like to see what it would take to get meson to support it.

It looks like the relevant code is here: https://github.com/mesonbuild/meson/blob/cb597adb0170a4818a4950ca9a4fe3f6d5a6c9d2/mesonbuild/environment.py#L106-L119

samurai doesn't support the `--version` option, and I am hesitant to add it for several reasons. However, it seems to me that meson shouldn't need to be doing this probing at all. The version is already checked by ninja itself using `ninja_required_version` in `build.ninja`, and if the user does not have it installed, they simply won't be able to start the build.

As far as I can tell, meson needs to know the ninja command for three purposes:
1. To support `ninja clean` as an alternative to `ninja -t clean`.
2. To support `ninja meson-scan-build` as an alternative to `scan-build ninja`.
3. To generate a compilation database during `build.ninja` generation.

Additionally, the version in `detect_ninja` is only used in one place in `run_tests.py`, presumably to make sure it supports `-w dupbuild=err`. However, earlier ninja versions will just fail because of an unsupported option.

One way to support samurai as well as any other future implementations would be to make a failure in `detect_ninja` just a warning, and in this case just omit a `clean` target. A failure in `ninja -t compdb` is already just a warning, so that could just be skipped if meson can't figure out the ninja command. Would this be acceptable?

**Comment by astavale** [comment]
Samurai looks good because it is built without any Python or C++ dependency. Obviously Meson requires Python, but Samurai seems a good step for anyone wanting to build a system without those dependencies.

I'm not a project maintainer, but as your question has been here a week, I'd thought I'd throw in my two cents if it offers any insight.

Meson likes to do a sanity check before the build. So letting the build fail when it comes to invoking Ninja is probably not a good idea. I imagine this is based on the usability of Meson. A build is set up with `meson setup my_build_dir` and that checks everything is present and correct before the build is run. At some point there will be a `meson build my_build_dir` command that would be an abstraction including `ninja -C my_build_dir`. Discussions are on going about the implementation. Meson abstracts away the details of compilers, backends, etc. to make a very easy to use build system. I'm guessing ninja isn't invoked in the setup stage apart from just checking the version.

I suggest looking at MSVC more closely. In the mesonbuild/backend/ directory vs2010backend.py has the `autodetect_vs_version(build)` function. Then vs2015backend.py and vs2017backend.py have a very limited number of overriding methods. I'm wondering if Samurai should take the same approach and inherit from ninjabackend.py. The version detection in environment.py can be refactored out to ninjabackend.py.

Also note that `meson --help` shows `--backend {ninja,vs,vs2010,vs2015,vs2017,xcode}`. So some consideration needs to go in to how Samurai fits in there.

I hope that helps a little.

**Comment by nirbheek** [comment]
An alternative to Ninja would definitely be interesting, but I think if you want to replace Ninja, you may want to go the same route as clang and [pkgconf](http://pkgconf.org/) by providing a wrapper that accepts the same command-line arguments as Ninja.

Then the roadmap is to reach a reasonable amount of feature parity and then start to fix some of the issues with Ninja. Such as:

1. Targets cannot have [multiple outputs in the dependency files
](https://github.com/ninja-build/ninja/issues/1184)
2. Does not [delete non-empty directories](https://github.com/ninja-build/ninja/issues/828) on platforms other than Windows
3. All command-lines are a single string, which leads to quoting hell
4. Requires `sh` on platforms other than Windows because of (3)
5. Support a 'non-human-readable' format for command-lines that have special characters in them (such as newlines or tabs, or anything else that is hard/impossible to escape). Current Meson has to use a python wrapper for this.
6. More that I can't remember right now.

To be fair, (3) is one valid solution to the problem that on Windows [`CreateProcess()`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682425(v=vs.85).aspx) is the native API and that accepts a single string as the command-line to run. `execve` and friends are [emulated by splitting that string](https://msdn.microsoft.com/en-us/library/windows/desktop/17w5ykft(v=vs.85).aspx).

However, I've come to believe that it may be more efficient and better for quoting/escaping if an array-like syntax for arguments was also supported, which on Windows would be converted to a single string command-line by doing the reverse of [CommandLineToArgvW()](https://msdn.microsoft.com/en-us/library/windows/desktop/bb776391(v=vs.85).aspx) before passing to `CreateProcess()`.

**Comment by nirbheek** [comment]
As an aside, not having MSVC dependency handling is a deal-breaker for many people that use Meson since reliable and consistent cross-platform behavior is one of the upsides of Meson.

**Comment by orbea** [comment]
As an update, `samurai` does support `--version` now so its possible to work around the version check issue and complete builds by symlinking `ninja` to `samu`. This is not ideal because the user may or may not have `ninja` installed.

**Comment by michaelforney** [comment]
I submitted pull request #4349, which adds `samu` to the list of commands to look for. This addresses the main issue.

However, I still think that meson should be more flexible here, at least with an option comparable to cmake's `-DCMAKE_MAKE_PROGRAM`.  This would let users specify a ninja command that might otherwise not be detected for any number of reasons (not in `PATH`, different name, or alternative implementation).

**Comment by orbea** [comment]
Agreed, I think the best way to address this is to have a list of known working implementations like `ninja` and `samu` while also offering an option to override it for whatever reason.

**Comment by jpakkane** [comment]
We specify other program overrides with envvars (`CC` et al). I guess we would use `NINJA` for this at least until "native file" support materialises.
